In [2]:
import pandas as pd
fold1_val = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/gold/fold-1/sentence_val_fold-1_word_format.csv")
fold2_val = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/gold/fold-2/sentence_val_fold-2_word_format.csv")
fold3_val = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/gold/fold-3/sentence_val_fold-3_word_format.csv")
fold4_val = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/gold/fold-4/sentence_val_fold-4_word_format.csv")
fold5_val = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/gold/fold-5/sentence_val_fold-5_word_format.csv")


In [3]:
len(fold1_val), len(fold2_val), len(fold3_val), len(fold4_val), len(fold5_val)


(124487, 122505, 125000, 123846, 124114)

In [4]:
fold2_val


post_id subreddit_id  sentence_id      words   labels
0       s1jpia     t5_2s23e            1          I        O
1       s1jpia     t5_2s23e            1      wrote        O
2       s1jpia     t5_2s23e            1       this        O
3       s1jpia     t5_2s23e            1          a        O
4       s1jpia     t5_2s23e            1        few        O
...        ...          ...          ...        ...      ...
122500  sivvy6     t5_2r876            7        now  per_exp
122501  sivvy6     t5_2r876            7       with  per_exp
122502  sivvy6     t5_2r876            7  Trikafta.  per_exp
122503  ri83g1     t5_2r876            6      Thank        O
122504  ri83g1     t5_2r876            6       you.        O

[122505 rows x 5 columns]

In [5]:
roberta_f2_e4 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/roberta-large_bs32_ga1_lr1e-05/fold-2/prediction_epoch-4.csv")
roberta_f3_e4 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/roberta-large_bs32_ga1_lr1e-05/fold-3/prediction_epoch-4.csv")
deberta_f2_e3 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/microsoft-deberta-large_bs32_ga1_lr1e-05/fold-2/prediction_epoch-3.csv")
deberta_f3_e3 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/microsoft-deberta-large_bs32_ga1_lr1e-05/fold-3/prediction_epoch-3.csv")


In [6]:
roberta_f2_e4


words labels
0               I      O
1           wrote      O
2            this      O
3               a      O
4             few      O
...           ...    ...
122500        now      O
122501       with      O
122502  Trikafta.      O
122503      Thank      O
122504       you.      O

[122505 rows x 2 columns]

In [7]:
deberta_f3_e3


words labels
0            When      O
1               I      O
2             was      O
3             17,      O
4              it      O
...           ...    ...
124995        the      O
124996  beginning      O
124997         of      O
124998        the      O
124999  pandemic.      O

[125000 rows x 2 columns]

In [8]:
fold2_val.loc[fold2_val["words"] != roberta_f2_e4["words"]]


Empty DataFrame
Columns: [post_id, subreddit_id, sentence_id, words, labels]
Index: []

In [9]:
fold3_val.loc[fold3_val["words"] != deberta_f3_e3["words"]]


Empty DataFrame
Columns: [post_id, subreddit_id, sentence_id, words, labels]
Index: []

In [10]:
# roberta 在 fold2 中對的數量

print("roberta_f2_e4 fold2 正確的有:" +
      str(len(fold2_val.loc[fold2_val['labels'] == roberta_f2_e4['labels']])))
print("deberta_f2_e3 fold2 正確的有:" +
      str(len(fold2_val.loc[fold2_val['labels'] == deberta_f2_e3['labels']])))


roberta_f2_e4 fold2 正確的有:87421
deberta_f2_e3 fold2 正確的有:88038


In [11]:
# roberta 在 fold2 中對的數量
print("roberta_f3_e4 fold3 正確的有:" +
      str(len(fold3_val.loc[fold3_val['labels'] == roberta_f3_e4['labels']])))
print("deberta_f3_e3 fold3 正確的有:" +
      str(len(fold3_val.loc[fold3_val['labels'] == deberta_f3_e3['labels']])))


roberta_f3_e4 fold3 正確的有:87664
deberta_f3_e3 fold3 正確的有:85706


In [12]:
corr_rob_f2 = len(
    fold2_val.loc[fold2_val['labels'] == roberta_f2_e4['labels']])
corr_deb_f2 = len(
    fold2_val.loc[fold2_val['labels'] == deberta_f2_e3['labels']])


In [13]:
# fold 2 m1=robert m2=deberta
m1c_m2c = fold2_val.loc[(fold2_val['labels'] == roberta_f2_e4["labels"]) & (
    fold2_val['labels'] == deberta_f2_e3["labels"])]
m1c_m2w = fold2_val.loc[(fold2_val['labels'] == roberta_f2_e4['labels']) & (
    fold2_val['labels'] != deberta_f2_e3['labels'])]
m1w_m2c = fold2_val.loc[(fold2_val['labels'] != roberta_f2_e4['labels']) & (
    fold2_val['labels'] == deberta_f2_e3['labels'])]
m1w_m2w = fold2_val.loc[(fold2_val['labels'] != roberta_f2_e4['labels']) & (
    fold2_val['labels'] != deberta_f2_e3['labels'])]


In [14]:
m1c_m2c


post_id subreddit_id  sentence_id  words    labels
0       s1jpia     t5_2s23e            1      I         O
1       s1jpia     t5_2s23e            1  wrote         O
2       s1jpia     t5_2s23e            1   this         O
3       s1jpia     t5_2s23e            1      a         O
4       s1jpia     t5_2s23e            1    few         O
...        ...          ...          ...    ...       ...
122462  sypylg     t5_2saq9            6    not  question
122463  sypylg     t5_2saq9            6    for  question
122464  sypylg     t5_2saq9            6  them?  question
122503  ri83g1     t5_2r876            6  Thank         O
122504  ri83g1     t5_2r876            6   you.         O

[79646 rows x 5 columns]

In [15]:
# fold2 mcnemar test model 1 and model 2
len(m1c_m2c), len(m1c_m2w), len(m1w_m2c), len(m1w_m2w)


(79646, 7775, 8392, 26692)

In [16]:
len(m1c_m2c) + len(m1c_m2w) + len(m1w_m2c) + len(m1w_m2w), len(fold2_val)


(122505, 122505)

In [17]:
import numpy as np
from mlxtend.evaluate import mcnemar

tb_fold2 = np.array([[len(m1c_m2c), len(m1c_m2w)],
                     [len(m1w_m2c), len(m1w_m2w)]])

chi2_f2, p_f2 = mcnemar(ary=tb_fold2, exact=True)

print(f"chi-squared of fold2: {chi2_f2}")
print(f"p-value of fold2: {p_f2}")


chi-squared of fold2: 7775
p-value of fold2: 1.2645142823545577e-06


In [18]:
# fold 2 m1=robert m2=deberta
m1c_m2c_f3 = fold3_val.loc[(fold3_val['labels'] == roberta_f3_e4["labels"]) & (
    fold3_val['labels'] == deberta_f3_e3["labels"])]
m1c_m2w_f3 = fold3_val.loc[(fold3_val['labels'] == roberta_f3_e4['labels']) & (
    fold3_val['labels'] != deberta_f3_e3['labels'])]
m1w_m2c_f3 = fold3_val.loc[(fold3_val['labels'] != roberta_f3_e4['labels']) & (
    fold3_val['labels'] == deberta_f3_e3['labels'])]
m1w_m2w_f3 = fold3_val.loc[(fold3_val['labels'] != roberta_f3_e4['labels']) & (
    fold3_val['labels'] != deberta_f3_e3['labels'])]


In [19]:
len(m1c_m2c_f3), len(m1c_m2w_f3), len(m1w_m2c_f3), len(m1w_m2w_f3)


(78708, 8956, 6998, 30338)

In [20]:
len(m1c_m2c_f3) + len(m1c_m2w_f3) + \
    len(m1w_m2c_f3) + len(m1w_m2w_f3), len(fold3_val)


(125000, 125000)

In [21]:
import numpy as np
from mlxtend.evaluate import mcnemar

tb_fold3 = np.array([[len(m1c_m2c_f3), len(m1c_m2w_f3)],
                     [len(m1w_m2c_f3), len(m1w_m2w_f3)]])

chi2_f3, p_f3 = mcnemar(ary=tb_fold3, exact=True)

print(f"chi-squared of fold3: {chi2_f3}")
print(f"p-value of fold3: {p_f3}")


chi-squared of fold3: 6998
p-value of fold3: 2.8251686573178394e-54


In [22]:
deberta_v3_f2_e4 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/microsoft-deberta-v3-large_bs32_ga1_lr1e-05/fold-2/prediction_epoch-4.csv")
deberta_v3_f3_e3 = pd.read_csv(
    "/workplace/yhcheng/SemEval_task8/workspace/simple_transformers/sentence_classification/eval/word_format/microsoft-deberta-v3-large_bs32_ga1_lr1e-05/fold-3/prediction_epoch-3.csv")


In [23]:
deberta_f3_e3


words labels
0            When      O
1               I      O
2             was      O
3             17,      O
4              it      O
...           ...    ...
124995        the      O
124996  beginning      O
124997         of      O
124998        the      O
124999  pandemic.      O

[125000 rows x 2 columns]

In [24]:
fold3_val.loc[deberta_v3_f3_e3['labels'] == deberta_f3_e3['labels']]


post_id subreddit_id  sentence_id      words labels
0       s1jpia     t5_2s23e            3       When      O
1       s1jpia     t5_2s23e            3          I      O
2       s1jpia     t5_2s23e            3        was      O
3       s1jpia     t5_2s23e            3        17,      O
4       s1jpia     t5_2s23e            3         it      O
...        ...          ...          ...        ...    ...
124995  rdbdw7     t5_2s1h9            7        the      O
124996  rdbdw7     t5_2s1h9            7  beginning      O
124997  rdbdw7     t5_2s1h9            7         of      O
124998  rdbdw7     t5_2s1h9            7        the      O
124999  rdbdw7     t5_2s1h9            7  pandemic.      O

[104185 rows x 5 columns]

In [25]:
# fold 2 m1=robert m2=deberta
m1c_m2c_v3 = fold2_val.loc[(fold2_val['labels'] == roberta_f2_e4["labels"]) & (
    fold2_val['labels'] == deberta_v3_f2_e4["labels"])]
m1c_m2w_v3 = fold2_val.loc[(fold2_val['labels'] == roberta_f2_e4['labels']) & (
    fold2_val['labels'] != deberta_v3_f2_e4['labels'])]
m1w_m2c_v3 = fold2_val.loc[(fold2_val['labels'] != roberta_f2_e4['labels']) & (
    fold2_val['labels'] == deberta_v3_f2_e4['labels'])]
m1w_m2w_v3 = fold2_val.loc[(fold2_val['labels'] != roberta_f2_e4['labels']) & (
    fold2_val['labels'] != deberta_v3_f2_e4['labels'])]


In [26]:
len(m1c_m2c_v3), len(m1c_m2w_v3), len(m1w_m2c_v3), len(m1w_m2w_v3)


(78444, 8977, 8115, 26969)

In [27]:
len(m1c_m2c_v3) + len(m1c_m2w_v3) + \
    len(m1w_m2c_v3) + len(m1w_m2w_v3), len(fold2_val)


(122505, 122505)

In [28]:
import numpy as np
from mlxtend.evaluate import mcnemar

tb_fold2_v2 = np.array([[len(m1c_m2c_v3), len(m1c_m2w_v3)],
                        [len(m1w_m2c_v3), len(m1w_m2w_v3)]])

chi2_f3_v3, p_f3_v3 = mcnemar(ary=tb_fold2_v2, exact=True)

print(f"chi-squared of fold3: {chi2_f3_v3}")
print(f"p-value of fold3: {p_f3_v3}")


chi-squared of fold3: 8115
p-value of fold3: 4.4839734931690804e-11


In [29]:
m1c_m2c_f3_v3 = fold3_val.loc[(fold3_val['labels'] == roberta_f3_e4["labels"]) & (
    fold3_val['labels'] == deberta_v3_f3_e3["labels"])]
m1c_m2w_f3_v3 = fold3_val.loc[(fold3_val['labels'] == roberta_f3_e4['labels']) & (
    fold3_val['labels'] != deberta_v3_f3_e3['labels'])]
m1w_m2c_f3_v3 = fold3_val.loc[(fold3_val['labels'] != roberta_f3_e4['labels']) & (
    fold3_val['labels'] == deberta_v3_f3_e3['labels'])]
m1w_m2w_f3_v3 = fold3_val.loc[(fold3_val['labels'] != roberta_f3_e4['labels']) & (
    fold3_val['labels'] != deberta_v3_f3_e3['labels'])]


In [30]:
len(m1c_m2c_f3_v3), len(m1c_m2w_f3_v3), len(m1w_m2c_f3_v3), len(m1w_m2w_f3_v3)


(80604, 7060, 8006, 29330)

In [31]:
len(m1c_m2c_f3_v3) + len(m1c_m2w_f3_v3) + \
    len(m1w_m2c_f3_v3) + len(m1w_m2w_f3_v3), len(fold3_val)


(125000, 125000)

In [34]:
print(m1w_m2w_f3_v3), print(roberta_f3_e4.loc[(fold3_val['labels'] != roberta_f3_e4['labels']) & (
    fold3_val['labels'] != deberta_v3_f3_e3['labels'])])


       post_id subreddit_id  sentence_id         words    labels
62      s1jpia     t5_2s23e           18          Even         O
63      s1jpia     t5_2s23e           18         today         O
64      s1jpia     t5_2s23e           18        almost         O
65      s1jpia     t5_2s23e           18        twenty         O
66      s1jpia     t5_2s23e           18         years         O
...        ...          ...          ...           ...       ...
124880  sivvy6     t5_2r876            6    considered  question
124881  sivvy6     t5_2r876            6       working  question
124882  sivvy6     t5_2r876            6            as  question
124883  sivvy6     t5_2r876            6    measurable  question
124884  sivvy6     t5_2r876            6  improvement.  question

[29330 rows x 5 columns]
               words   labels
62              Even  per_exp
63             today  per_exp
64            almost  per_exp
65            twenty  per_exp
66             years  per_exp
...           

(None, None)

In [32]:
import numpy as np
from mlxtend.evaluate import mcnemar

tb_fold3_v3 = np.array([[len(m1c_m2c_f3_v3), len(m1c_m2w_f3_v3)],
                        [len(m1w_m2c_f3_v3), len(m1w_m2w_f3_v3)]])

chi2_f3_v3_2, p_f3_v3_2 = mcnemar(ary=tb_fold3_v3, exact=True)
print(tb_fold3_v3)
print(f"chi-squared of fold3: {chi2_f3_v3_2}")
print(f"p-value of fold3: {p_f3_v3_2}")


[[80604  7060]
 [ 8006 29330]]
chi-squared of fold3: 7060
p-value of fold3: 1.3452170712337588e-14


# Calculating the p-value of deberta-large & roberta-large on data split in fold2 and fold3

## model:

m1: roberta-large  
m2: deberta-large-v1

- fold2 p-value: 8.427527424066525e-35
- fold2 chi-square: 6957

- fold3 p-value: 2.8251686573178394e-54
- fold3 chi-square: 6998

fold2 labels total: (122505) 在資料分割折數 2 時
| | m2 correct | m2 wrong|
|-----|-----|-----|
| m1 correct| 80464| 6957 |
| m1 wrong| 8486 | 26598|

fold3 labels total: (125000) 在資料分割折數 3 時
| | m2 correct | m2 wrong|
|-----|-----|-----|
| m1 correct| 78708| 8956 |
| m1 wrong| 6998 | 30338|

---

# Calculating the p-value of deberta-large-v3 & roberta-large on data split in fold2 and fold3

## model:

m1: roberta-large  
 m2: deberta-large-v3

- fold2 p-value: 4.4839734931690804e-11
- fold2 chi-square: 8115

- fold3 p-value: 1.3452170712337588e-14
- fold3 chi-square: 7060

fold2 labels total: (122505) 在資料分割折數 2 時

|            | m2 correct | m2 wrong |
| ---------- | ---------- | -------- |
| m1 correct | 78444      | 8977     |
| m1 wrong   | 8115       | 26969    |

fold3 labels total: (125000) 在資料分割折數 3 時
| | m2 correct | m2 wrong|
|-----|-----|-----|
| m1 correct| 80604| 7060 |
| m1 wrong| 8006 | 29330|
